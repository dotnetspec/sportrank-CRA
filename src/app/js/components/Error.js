import React, { Component } from 'react';
import { Grid, Row, Col } from 'react-bootstrap';
//import Grid from 'react-bootstrap/Grid'
//import PageHeader from 'react-bootstrap/PageHeader'

/**
 * Class representing the home page rendering
 *
 * @extends React.Component
 */
class Error extends Component {

  //#region Constructor
  // constructor(props) {
  //   super(props);
  // }
  //#endregion

  //#region React lifecycle events
  render() {

    const metaMaskPossible = (this.props.error.message.indexOf('Internal JSON-RPC error') > 0 || this.props.error.message.indexOf('Failed to fetch') > 0);

    return (
      <Grid>
        <Row>
          <Col xs={12}>
            <h2>
              Whoopsieee <small>Something went wrong</small>
            </h2>
            {metaMaskPossible ?
              <React.Fragment>
                <h3>Metamask error?</h3>
                <p>It appears you might be using metamask. Have you signed in and are you on the right network?</p>
              </React.Fragment>
              :
              ''
            }
            <h3>Error details</h3>
            <pre>{this.props.error.message}<br />
            {this.props.error.stack}
            { this.props.error.source ?
              <React.Fragment>
                <br/>Source: { this.props.error.source }
              </React.Fragment>
              :
              ''
            }</pre>
          </Col>
        </Row>
      </Grid>
    );
  }
  //#endregion
}

export default Error
