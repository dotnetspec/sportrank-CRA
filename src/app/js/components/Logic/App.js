import Header from '../UI/Header'
import Main from './Main'
import React, {
  useState,
  useEffect
} from 'react';
import {
  withRouter
} from 'react-router-dom';
import JSONops from './JSONops';
import {
  _loadsetJSONData,
  getNewRankId,
  getDefaultRankingList
} from '../SideEffects/io/Jsonio';
import {
  _loadExternalBalance
  //,
  // _mapCurrentUserAccounts,
  // mapTheAccounts
} from '../SideEffects/io/web3io';
import web3 from '../../../../web3';
import DSportRank from '../../../../ABIaddress';
import ChangeState from './ChangeState';

/**
 * Functional component representing the highest order component. Any user
 * updates in child components should trigger an event in this
 * class so that the current user details can be re-fetched from
 * the contract and propagated to all children that rely on it
 *
 * @extends React.Component
 */
export function App({
  props
}) {
  const [user, setUser] = useState({});
  const [userName, setUserName] = useState('');
  const [account, setAccount] = useState('');
  const [error, setError] = useState({});
  const [userAccounts, setuserAccounts] = useState([])
  const [balance, setBalance] = useState(0)
  const [data, setdata] = useState([])
  const [updatedExtAcctBalCB, setupdatedExtAcctBalCB] = useState(0)
  const [isLoading, setIsLoading] = useState(true);
  const [userRankingLists] = useState([]);
  const [isUserInJson, setIsUserInJson] = useState(false);
  const [isLoadingJSON, setisLoadingJSON] = useState(true);
  const [isCurrentUserActive, setIsCurrentUserActive] = useState(false);
  const [isCurrentUserActiveCB] = useState(false);
  const [isRankingIDInvalid] = useState(false);
  const [newrankId, setnewrankId] = useState('');
  const [rankingListData, setrankingListData] = useState([]);
  const [viewingOnlyCB, setviewingOnlyCB] = useState(true);
  const [contactno, setcontactno] = useState('');
  const [email, setemail] = useState('');
  const [description, setdescription] = useState({});
  const [specificRankingOptionBtns, setspecificRankingOptionBtns] = useState(false);
  const [rank, setrank] = useState('1');
  const [ranknameHasChanged, setranknameHasChanged] = useState(false);
  const [resultInfoForDisplay, setResultInfoForDisplay] = useState('');
  //REVIEW: set as a global var. Perhaps change to environment var ?:
  const rankingDefault = '5c36f5422c87fa27306acb52';

  function _loadExternalBalance_callback(externalbalance) {
    if (externalbalance !== undefined) {
      setupdatedExtAcctBalCB(externalbalance);
    }
  }

  //Below appears to be relevant to user events not e.g. callbacks that fetch data
  const setspecificRankingOptionBtnsCB = () => {
    setspecificRankingOptionBtns(true);
  }
  const handleListAllChildClick = () => {
    setspecificRankingOptionBtns(false);
  }
  //cb from GlobalRankings.js to set the rank state as view only
  //REVEIW: is this necessary? wasn't working until noticed it ...
  const setOnCallbackviewingOnlyCB = (viewingOnlyCBval) => {
    setviewingOnlyCB(viewingOnlyCBval);
  }
  //cb from PlayerStatusBtn.js to set the state of the button
  const setOnCallbackisCurrentUserActiveCB = (btnState) => {
    setIsCurrentUserActive(btnState);
  }
  //in GlobalRankingViewBtn
  const setnewrankIdCB = (newrankIdtxt) => {
    setnewrankId(newrankIdtxt);
    setranknameHasChanged(true);
    setIsLoading(false);
    _loadsetJSONData(newrankIdtxt, setrankingJSONdataCB);
  }
  const setResultInfoForDisplayCB = (text) => {
    console.log('setResultInfoForDisplayCB in app.js', text);
    setResultInfoForDisplay(text);
  }
  //cb from createuser.js to set the username
  //in time for getNewRankingID() to put it in the json
  const setuserNameCB = (name) => {
    setUserName(name);
    //this was originally
    //may need to account for this
    //setUser(name);
  }
  const setcontactNoCB = (number) => {
    setcontactno(number);
  }
  const setemailCB = (oppoEmailTxt) => {
    setemail(oppoEmailTxt);
  }
  const setuserDescCB = (userObj, description) => {
    userObj.description = description;
    setUser(userObj);
  }

  const setuserCB = (userObj, name, contactno, email, description) => {
    userObj.username = name;
    userObj.contactno = contactno;
    userObj.email = email;
    userObj.description = description;
    setUser(userObj);
  }

  const setrankingJSONdataCB = (datatoSet) => {
    setdata(datatoSet);
    setisLoadingJSON(false);
    setIsUserInJson(JSONops.isPlayerListedInJSON(data, user.username));
    setrank(JSONops._getUserValue(data, user.username, "RANK"));
    setIsCurrentUserActive(JSONops._getUserValue(data, user.username, "ACTIVE"));
  }
  //#endregion
  //#region Helper methods
  //REVIEW: Possible to getUserRank in App.js (and set state) rather than Home.js?
  //currently no - problem is waiting for username to check against ran
  /**
   * Sets the App state error and redirects the user to the error page
   *
   * @param {Error} err - error encountered
   */
  const _onError = (err, source) => {
    if (source) err.source = source;
    //setState({ error: err });
    setError(err);
    props.history.push('/whoopsie');
  }
  //#endregion

  const processStateAfter_loadCurrentUserAccounts = (userAcctArr) => {
    console.log('userAcctArr before clean', userAcctArr)
    userAcctArr = ChangeState.cleanUpUserSRAccountData(userAcctArr);
    console.log('userAcctArr after clean', userAcctArr)
    if (userAcctArr) {
      setuserAccounts(userAcctArr);
      //console.log('userAcctArr[0].userAccount', userAcctArr[0])
      //REVIEW: sort out this duplication
      setAccount(userAcctArr[0]);
      setUser(userAcctArr[0])
      setError(userAcctArr[0].error);
      setUserName(userAcctArr[0].username);
      console.log(userAcctArr[0])
      if (userAcctArr[0].data !== undefined) {
        setIsUserInJson(JSONops.isPlayerListedInJSON(userAcctArr[0].data, userAcctArr[0].username));
        setIsCurrentUserActive(JSONops._getUserValue(userAcctArr[0].data, userAcctArr[0].username, "ACTIVE"));
      } else {
        setIsUserInJson(false);
      }
      setnewrankId(userAcctArr[0].newrankId);
      setcontactno(userAcctArr[0].contactno);
      setemail(userAcctArr[0].email);
      setdescription(userAcctArr[0].description);
      setBalance(userAcctArr[0].balance);
      setviewingOnlyCB(true);
    } else {
      console.log('user undefined')
    }
  }
  //#region React lifecycle events
  useEffect(() => {
    setIsLoading(true);
    async function fetchData() {
      //from the Blockchain via web3io:
      //await _loadExternalBalance(_loadExternalBalance_callback);

      await getDefaultRankingList(rankingDefault, getDefaultRankingList_callback);
      async function getDefaultRankingList_callback(json) {
        setrankingListData(json);
      }

      const mapTheAccounts = async (accountsArray) => {
        web3.eth.getAccounts()
        .then(function(expectingAResult){
            return expectingAResult;
        })
        .then(async function(accountsArray){
          //based on: https://flaviocopes.com/javascript-async-await-array-map/
          //get the hashes by async
          //each separate piece of async data needs these parts ...
          const functionWithPromise = item => { //a function that returns a promise
            return Promise.resolve(DSportRank.methods.owners(item).call())
          }

          const anAsyncFunction = async item => {
            return await functionWithPromise(item)
          }
          //wait for the async hash calls to resolve for all accounts
          const getData = async () => {
            return await Promise.all(accountsArray.map(item => anAsyncFunction(item)))
          }
          const data = await getData()
          //each separate piece of async data needs these parts (above)  ...

          //use the hashes to get the user data from contract by async
          const functionGetUserDataWithPromise = item => { //a function that returns a promise
            return Promise.resolve(DSportRank.methods.users(item).call());
          }

          const anAsyncFunctionToGetUserData = async item => {
            return await functionGetUserDataWithPromise(item)
          }
          //wait for the async hash calls to resolve for all users
          const getUserData = async () => {
            return await Promise.all(data.map(item => anAsyncFunctionToGetUserData(item)))
          }

          const userdata = await getUserData()
          console.log('user data in the list', userdata)

          const functionWithPromiseToGetBal = item => { //a function that returns a promise
            return Promise.resolve(web3.eth.getBalance(item.owner));
          }

          const anAsyncFunctionToGetBal = async item => {
            return await functionWithPromiseToGetBal(item)
          }
          //wait for the async hash calls to resolve for all accounts
          const getDataWithBal = async () => {
            return await Promise.all(userdata.map(item => anAsyncFunctionToGetBal(item)))
          }
          const userdataWithBal = await getDataWithBal()

          //var numbers = [4, 9, 16, 25];
          //var userdataWithBalArr = userdata.map(item => web3.eth.getBalance('0xAC5491BB066c98fec13046928a78761c0B1E5603'));
          console.log('userdataWithBal', userdataWithBal)
          return userdata;
        }).then(function(resolvedUserData){
                  processStateAfter_loadCurrentUserAccounts(resolvedUserData);
              }).catch(function(error) {
                   console.log('error is:', error)
              }).then(function() {
                   console.log('in the last then')
              });
            }
        await mapTheAccounts();

        setIsLoading(false);
    }
    fetchData();
  }, []); // Or [someId] (sent as a param to a function) if effect needs props or state (apparently)

  if (!isLoading) {
    console.log('userName', userName)
    return ( <
      div >
      <
      Header data-testid = 'header'
      user = {
        user
      }
      username = {
        userName
      }
      account = {
        account
      }
      description = {
        description
      }
      setuserDescCB = {
        setuserDescCB
      }
      userAccounts = {
        userAccounts
      }
      balance = {
        balance
      }
      error = {
        error
      }
      isCurrentUserActive = {
        isCurrentUserActive
      }
      isCurrentUserActiveCB = {
        isCurrentUserActiveCB
      }
      setOnCallbackisCurrentUserActiveCB = {
        setOnCallbackisCurrentUserActiveCB
      }
      setspecificRankingOptionBtnsCB = {
        setspecificRankingOptionBtnsCB
      }
      onListAllChildClick = {
        handleListAllChildClick
      }
      specificRankingOptionBtns = {
        specificRankingOptionBtns
      }
      // onAfterUserUpdate = {
      //   mapTheAccounts
      // }
      onError = {
        _onError
      }
      rankingJSONdata = {
        data
      }
      rankingListJSONdata = {
        rankingListData
      }
      updatedExtAcctBalCB = {
        updatedExtAcctBalCB
      }
      usersRankingLists = {
        userRankingLists
      }
      isUserInJson = {
        isUserInJson
      }
      rankingDefault = {
        rankingDefault
      }
      newrankId = {
        newrankId
      }
      setnewrankId = {
        setnewrankId
      }
      resultInfoForDisplay = {
        resultInfoForDisplay
      }
      setResultInfoForDisplayCB = {
        setResultInfoForDisplayCB
      }
      />

      <
      Main data-testid = 'main'
      user = {
        user
      }
      setuserCB = {
        setuserCB
      }
      username = {
        userName
      }
      setuserNameCB = {
        setuserNameCB
      }
      contactno = {
        contactno
      }
      setcontactNoCB = {
        setcontactNoCB
      }
      email = {
        email
      }
      setemailCB = {
        setemailCB
      }
      description = {
        description
      }
      setuserDescCB = {
        setuserDescCB
      }
      resultInfoForDisplay = {
        resultInfoForDisplay
      }
      setResultInfoForDisplayCB = {
        setResultInfoForDisplayCB
      }
      account = {
        account
      }
      userAccounts = {
        userAccounts
      }
      error = {
        error
      }
      setspecificRankingOptionBtnsCB = {
        setspecificRankingOptionBtnsCB
      }
      specificRankingOptionBtns = {
        specificRankingOptionBtns
      }

      // onAfterUserUpdate = {
      //   _mapCurrentUserAccounts
      // }
      onError = {
        _onError
      }
      rankingJSONdata = {
        data
      }
      setrankingJSONdataCB = {
        setrankingJSONdataCB
      }
      rankingListJSONdata = {
        rankingListData
      }
      rank = {
        rank
      }
      ranknameHasChanged = {
        ranknameHasChanged
      }
      isCurrentUserActive = {
        isCurrentUserActive
      }
      setOnCallbackisCurrentUserActiveCB = {
        setOnCallbackisCurrentUserActiveCB
      }
      isRankingIDInvalid = {
        isRankingIDInvalid
      }
      newrankId = {
        newrankId
      }
      setnewrankIdCB = {
        setnewrankIdCB
      }
      getNewRankingID = {
        getNewRankId
      }
      rankingDefault = {
        rankingDefault
      }
      viewingOnlyCB = {
        viewingOnlyCB
      }
      setviewingOnlyCB = {
        setOnCallbackviewingOnlyCB
      }
      isUserInJson = {
        isUserInJson
      }
      loadingJSON = {
        isLoadingJSON
      }
      />

      <
      /div>
    );
  } else {
    return ( <
      div data-testid = 'loading' > Loading... < /div>
    );
  }
  //}
  //#endregion
}

export default withRouter(App)
