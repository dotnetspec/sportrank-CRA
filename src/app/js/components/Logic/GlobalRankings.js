//import { Switch, Route } from 'react-router-dom';
import { BootstrapTable, TableHeaderColumn } from 'react-bootstrap-table';
//import { Grid, Row, Col, PageHeader, Image, Modal, Navbar, ButtonToolbar, Dropdown, DropdownItem, Overlay, Tooltip, Button, FormGroup, ControlLabel, FormControl, HelpBlock } from 'react-bootstrap';
import { Grid, Row } from 'react-bootstrap';
import { isEmpty } from '../../utils';
import GlobalRankingViewBtn from '../UI/buttons/GlobalRankingViewBtn';
import GlobalRankingJoinBtn from '../UI/buttons/GlobalRankingJoinBtn';
import React, { Component } from 'react';

// const selectRowPropAfterClickRow = {
//   selectedRankingId: ''
// };

/**
 * Class handling the global ranking selection list
 *
 * User clicks on a ranking to display and use it
 *
 * @extends React.Component
 */

class GlobalRankings extends Component {

  //#region Constructor
  constructor(props){
    super(props);
    this.state = {
      data:''
      // ,
      // //error currently not used, so set to false
      // error: false
    }
    this.tablesortoptions = {
     defaultSortName: 'RANKINGNAME',  // default sort column name
     defaultSortOrder: 'asc'  // default sort order
   };
  }
  //#endregion


   rankingViewButton(cell, row, enumObject, rowIndex) {
       return (
            <GlobalRankingViewBtn
            cell={cell} row={row} rowIndex={rowIndex}
            setspecificRankingOptionBtnsCB={this.props.setspecificRankingOptionBtnsCB}
            onAfterUserUpdate={this.props.onAfterUserUpdate}
            history={this.props.history}
            username={this.props.username}
            setnewrankIdCB={this.props.setnewrankIdCB}
            viewingOnlyCB={this.props.viewingOnlyCB}
            newrankId={this.props.newrankId}
            setviewingOnlyCB={this.props.setviewingOnlyCB}
            />
       )
    }

    rankingJoinButton(cell, row, rowIndex){
      return (
           <GlobalRankingJoinBtn
           cell={cell} row={row} rowIndex={rowIndex}
           setspecificRankingOptionBtnsCB={this.props.setspecificRankingOptionBtnsCB}
           onAfterUserUpdate={this.props.onAfterUserUpdate}
           history={this.props.history}
           username={this.props.username}
           setnewrankIdCB={this.props.setnewrankIdCB}
           newrankId={this.props.newrankId}
           viewingOnlyCB={this.props.viewingOnlyCB}
           setviewingOnlyCB={this.props.setviewingOnlyCB}
           setrankingJSONdataCB={this.props.setrankingJSONdataCB}
           account={this.props.account}
           setOnCallbackisCurrentUserActiveCB={this.props.setOnCallbackisCurrentUserActiveCB}
           />
      )
    }
//#region React lifecycle events
//QUESTION: why does componentDidMount not have the data from this.props.rankingJSONdata
//when it clearly gets passed to Home.js?

  globalBSTableDisplay(){
      //NB: to enable non-test jsonbin.io data use the following as a property of
      //data={this.props.rankingListJSONdata}
      //original test: data={this.state.data}
      return (
          <BootstrapTable  options={ this.tablesortoptions } data={this.props.rankingListJSONdata}
          className='bstable'
          >
                <TableHeaderColumn isKey dataField='RANKINGID'
                hidden >
                  RANKINGID
                </TableHeaderColumn>

                <TableHeaderColumn  dataField='RANKINGNAME'
                filter={ { type: 'TextFilter', defaultValue: '' } }
                >
                  Ranking Name (Filter)
                </TableHeaderColumn>

                <TableHeaderColumn dataField='RANKINGDESC'
                filter={ { type: 'TextFilter',  defaultValue: '' } }
                >
                 Ranking Description (Filter)
                </TableHeaderColumn>

                <TableHeaderColumn
                dataField='viewbtn'
                dataFormat={this.rankingViewButton.bind(this)}
                >
                View
                </TableHeaderColumn>
                <TableHeaderColumn
                dataField='joinbtn'
                dataFormat={this.rankingJoinButton.bind(this)}
              >
                Join
                </TableHeaderColumn>

                <TableHeaderColumn dataField='ACTIVE'
                filter={ { type: 'TextFilter', defaultValue: 'true' } }
                hidden>
                  Active?
                </TableHeaderColumn>
              </BootstrapTable>
          )
        }

        componentDidMount() {
          //if there's no user re-direct to create
          if(this.props.user === ''){
            this.props.history.push('/create');
          }
        }



  //REVIEW: Home page may be unnecessarily re-rendering with this approach to passing props
  //but need to pass the username and display it as a greeting and to link account with json data
  //this.props.user[1] is a quick way (not object.keys) to access the array
  //if need it get   {React.version} by adding to return() below: (currently 16.6.3)
  render () {
    return (
      <div>
      {isEmpty(this.props.error) ? null : <span className='error'>Please note: there was an ERROR!</span>}
      <Grid>
        <Row>
          {this.globalBSTableDisplay()}
        </Row>
      </Grid>
      </div>
    )
  }
  //#endregion
}

export default GlobalRankings
