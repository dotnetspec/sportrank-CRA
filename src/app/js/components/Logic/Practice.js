// Test.js
//import React, { Component } from 'react';
import React from 'react';

// class Practice extends Component {
//   componentDidMount() {
//     this.func();
//   }
//   func = () => {
//     // noop
//   };
//   render() {
//     return null;
//   }
// }
//
// export default Practice

export const Practice = () => {
  //'value' here is equivalent of a state var in an object component
  const [value, setValue] = React.useState('initial state');

  const onChange = event => setValue(event.target.value);

  return (
    <div>
      <h1>Hello React Function Component!</h1>

      <input value={value} type="text" onChange={onChange} />

      <p>{value}</p>
    </div>
  );
};

export const setValInPractice = (newState) => ({
  value: newState.value,
});
