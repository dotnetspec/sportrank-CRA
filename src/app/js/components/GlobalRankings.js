//import { Switch, Route } from 'react-router-dom';
import { BootstrapTable, TableHeaderColumn } from 'react-bootstrap-table';
//import { Grid, Row, Col, PageHeader, Image, Modal, Navbar, ButtonToolbar, Dropdown, DropdownItem, Overlay, Tooltip, Button, FormGroup, ControlLabel, FormControl, HelpBlock } from 'react-bootstrap';
import { Grid, Row } from 'react-bootstrap';


//import PropsRoute from './PropsRoute';
import Home from './Home';
//import Userchallenges from './Userchallenges';
// import CreateUser from './CreateUser';
// import UpdateUser from './UpdateUser';
// import DeactivatePlayer from './DeactivatePlayer';
//import Error from './Error';
import React, { Component } from 'react';
// import JSONops from './JSONops'
// import CreateNewRanking from './CreateNewRanking';
//import UserRankings from './UserRankings'
import {newrankIdCB, viewingOnlyCB} from './App'
//import {saveJson, loadJson} from '../lib/service'
//import Grid from 'react-bootstrap/Grid'

const selectRowPropAfterClickRow = {
  selectedRankingId: ''
};

/**
 * Class handling the global ranking selection list
 *
 * User clicks on a ranking to display and use it
 *
 * @extends React.Component
 */

class GlobalRankings extends Component {

  //#region Constructor
  constructor(props){
    super(props);
    this.state = {
      data:''
      // ,
      // error: false
    }
    this.tablesortoptions = {
     defaultSortName: 'RANKINGNAME',  // default sort column name
     defaultSortOrder: 'asc'  // default sort order
   };
   console.log('in GlobalRankings')
  }
  //#endregion


  //REVIEW: change name to onClickRankingJoinSelected?
  onClickRankingSelected(cell, row, rowIndex){
    selectRowPropAfterClickRow.selectedRankingId = `${row['RANKINGID']}`;
    console.log('selectRowPropAfterClickRow.selectedRankingId', selectRowPropAfterClickRow.selectedRankingId)
    newrankIdCB(selectRowPropAfterClickRow.selectedRankingId);
    viewingOnlyCB(false);
    this.props.onAfterUserUpdate();
    //if joining and not yet a member of the ranking home will add the new player to the bottom
    //of the rankings in the selected ladder
    this.props.history.push('/home/@' + this.props.user.username);
    //this.openResultModal();
   }

   onClickRankingViewSelected(cell, row, rowIndex){
     //console.log('Product #', rowIndex);
     selectRowPropAfterClickRow.selectedRankingId = `${row['RANKINGID']}`;
     console.log('selectRowPropAfterClickRow.selectedRankingId', selectRowPropAfterClickRow.selectedRankingId)
     newrankIdCB(selectRowPropAfterClickRow.selectedRankingId);
     viewingOnlyCB(true);
     this.props.onAfterUserUpdate();
     this.props.history.push('/home/@' + this.props.user.username);
     //this.openResultModal();
    }

  rankingButton(cell, row, enumObject, rowIndex) {
      return (
         <button
            bsstyle="primary"
            //type="button"
            onClick={() =>
            this.onClickRankingSelected(cell, row, rowIndex)}
         >
         Join
         </button>
      )
   }

   rankingViewButton(cell, row, enumObject, rowIndex) {
       return (
          <button
             bsstyle="primary"
             //type="button"
             onClick={() =>
             this.onClickRankingViewSelected(cell, row, rowIndex)}
          >
          View
          </button>
       )
    }

//#region React lifecycle events
//QUESTION: why does componentDidMount not have the data from this.props.rankingJSONdata
//when it clearly gets passed to Home.js?
  componentDidMount() {
    //this.setState({error:false});
    //NB: loadJson is currently a testing function
    //and the 'data' comes from the test stub
    // loadJson()
    // .then (({data}) => this.setState({data:data}))
    // .catch(() => this.setState({error:true}));
  }

  globalBSTableDisplay(){
      //if (this.props.rankingJSONdata[0] === null && this.props.user.username === null){
      // if (JSONops.isJSONEmpty(this.props.rankingJSONdata) && this.props.user.username === null){
      //
      console.log('inside globalBSTableDisplay');


      //   this.props.history.push('/create');
      //   return null;
      //   //(<div>No Data To Display - Please select an account (top right) to create a player</div>);
      // } else {


      //NB: to enable non-test jsonbin.io data use the following as a property of
      //data={this.props.rankingListJSONdata}
      //original test: data={this.state.data}
      return (
        <BootstrapTable  options={ this.tablesortoptions } data={this.props.rankingListJSONdata}
        className='bstable'
        >
              <TableHeaderColumn isKey dataField='RANKINGID'
              hidden >
                RANKINGID
              </TableHeaderColumn>

              <TableHeaderColumn  dataField='RANKINGNAME'
              filter={ { type: 'TextFilter', defaultValue: '' } }
              >
                Ranking Name (Filter)
              </TableHeaderColumn>

              <TableHeaderColumn dataField='RANKINGDESC'
              filter={ { type: 'TextFilter',  defaultValue: '' } }
              >
               Ranking Description (Filter)
              </TableHeaderColumn>

              <TableHeaderColumn
              dataField='button'
              dataFormat={this.rankingViewButton.bind(this)}
            >
              View
              </TableHeaderColumn>

              <TableHeaderColumn
              dataField='button'
              dataFormat={this.rankingButton.bind(this)}
            >
              Join
              </TableHeaderColumn>

              <TableHeaderColumn dataField='ACTIVE'
              filter={ { type: 'TextFilter', defaultValue: 'true' } }
              hidden>
                Active?
              </TableHeaderColumn>

            </BootstrapTable>
          )
        }


  //REVIEW: Home page may be unnecessarily re-rendering with this approach to passing props
  //but need to pass the username and display it as a greeting and to link account with json data
  //this.props.user[1] is a quick way (not object.keys) to access the array
  //if need it get   {React.version} by adding to return() below: (currently 16.6.3)
  render () {
      console.log('b4 render globalBSTableDisplay with rankingListJSONdata', this.props.rankingListJSONdata)

    return (
      <div>
      {this.props.error ? <span className='error'>Oh no!</span> : null}
      <Grid>
        <Row>
          {this.globalBSTableDisplay()}
        </Row>
      </Grid>
      </div>
    )
  }
  //#endregion
}

export default GlobalRankings
