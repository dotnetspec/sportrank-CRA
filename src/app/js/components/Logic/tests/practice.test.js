import React, {useState} from 'react'
import { BrowserRouter, MemoryRouter } from 'react-router-dom'
import App  from '../App'
import PropsRoute  from '../PropsRoute'
import Error from '../Error'
import GlobalRankings from '../GlobalRankings'
//import { expect } from 'chai';
//import utils from '../../../utils'
//import * as jsonops from '../JSONops'
//import JSONops from '../JSONops'
//import * as app from "../App";
import Home from "../Home";
import Main from "../Main";
//import renderer from 'react-test-renderer'
import { shallow, mount } from 'enzyme';
import { isJSONEmpty } from '../JSONops'; // isJSONEmpty here is a mock function
import * as renderer from 'react-test-renderer'
//import * as pleaseMockMeFuncts from '../Practise'
//auto mock all the exported functions in utils.js
//jest.mock("./JSONops");
//const app = shallow(<App />)
   //const instance = jsonops.instance()
//const addMock = jest.spyOn(JSONops, "isJSONEmpty");


// practice.spec.js
//curly braces because, as a functional component, it is now a 'named' import
import { Practice, setValInPractice } from '../Practice';

// fit('PRACTICE - test a functional component', () => {
//   shallow(<Practice />);
//   expect(Practice.prototype.componentDidMount).toHaveBeenCalled();
// });

xit('should change the state of Practice', () => {
  //const component = renderer.create(<Practice />)
  // const instance = component.getInstance()
  // global.jestExpect(instance.state.value).to.equal('new state');
  //const PracticeFunct = new Practice();
  //const [value, setValue] = useState("");
  expect(Practice.state.value).to.equal('new state');
   //const state = { value: 'new state' };
   //const newState = setValInPractice(state);
   const [value, setValue] = useState("new state");
   //chai expect
   //expect(newState.value).to.equal('new state');
   expect(value).to.equal('new state');
 });

//taken from playerstatusbtn.test.js
xit('PRACTICE - test a functional component', () => {
  //const historyMock = { push: jest.fn() };
  //const component = renderer.create(<Practice history={historyMock}/>)
  const component = renderer.create(<Practice />)
  //console.log(component.instance().state);
  //const component = mount(shallow(<Practice />).get(0))
  expect(component.toJSON()).toMatchSnapshot()
  // getInstance is returning the `this` object you have in your component
  // meaning anything accessible on `this` inside your component
  // can be accessed on getInstance, including props!
  const instance = component.getInstance()

  global.jestExpect(instance.state).toMatchSnapshot('initial state')
  instance.onChange({ target: { value: 'updated stateu' } })
  global.jestExpect(instance.state).toMatchSnapshot('updated state')
});


// it('check the onChange callback', () => {const onChange = jest.fn(),
//   props = { value: '20.01.2018', onChange},
//   DateInputComponent = mount(<DateInput {...props} />).find('input');
//   DateInputComponent.simulate('change', { target: {value: moment('2018-01-22')} });
//   expect(onChange).toHaveBeenCalledWith('22.01.2018');
// });

xit('PRACTICE - invokes `componentDidMount` when mounted', () => {
  jest.spyOn(Practice.prototype, 'componentDidMount');
  shallow(<Practice />);
  expect(Practice.prototype.componentDidMount).toHaveBeenCalled();
  Practice.prototype.componentDidMount.mockRestore();
});

xit('Main - invokes `componentDidMount` when mounted', () => {
  jest.spyOn(Main.prototype, 'componentDidMount');
  shallow(<Main />);
  expect(Main.prototype.componentDidMount).toHaveBeenCalled();
  Main.prototype.componentDidMount.mockRestore();
});


xit("trying jest.mock properties with Practise", () => {
const myMock = jest.fn();

const a = new myMock();
const b = {};
const bound = myMock.bind(b);
bound();

console.log(myMock.mock.instances);
});


xit("trying jest.mock with Practise", () => {
  jest.mock('../Practise'); // this will mock all the exported members in ./Service
  //pleaseMockMeFuncts.get.mockResolvedValue(resp);
  //jest.SpyOn(pleaseMockMeFuncts, 'pleaseMockMe');
  // pleaseMockMeFuncts.pleaseMockMe(); // calling mock function
  // expect(pleaseMockMeFuncts.pleaseMockMe).toHaveBeenCalledTimes(1);
  //pleaseMockMe(); // calling mock function
  //expect(pleaseMockMe).toHaveBeenCalledTimes(1);

  // The function was called exactly once
//expect(pleaseMockMeFuncts.pleaseMockMe.mock.calls.length).toBe(1);
});

  xit("trying jest.mock with App", () => {
    // const data = [
    // {
    //   "RANKINGNAME":"testRa1","RANKINGDESC":"testRank","ACTIVE":true,"RANKINGID":"5c6a7cf5a83a2931773847b8"
    // }];
  //app.doAdd(1, 2);
  //getIdNoFromJsonbinResponse
  //app._loadsetJSONData_callback(data);
  //App.newrankIdCB(1234567);
  //app.newrankIdCB(1234567);
  // var user = {};
  // user.username = 'player1';
  // const app = shallow(<Home user={user}/>)
  Main.componentDidMount();
  //JSONops.isPlayerListedInJSON();
  //expect(addMock).toHaveBeenCalledWith();
});

// it("calls math.subtract", () => {
//   app.doSubtract(1, 2);
//   expect(math.subtract).toHaveBeenCalledWith(1, 2);
// });



//REVIEW: practice code:
xit('simple mock function', () => {

  const mockFunct = jest.fn(() => 2);

    expect(mockFunct(1,2)).toBe(2);
    //expect(mockFunct(1,2)).toBe(2);
    expect(mockFunct).toHaveBeenCalledTimes(1);
    expect(mockFunct).toHaveBeenCalledWith(1,2);

  });

  //simple callback function test
  const doAdd = (a, b, callback) => {
      callback(a + b);
    };

xit("calls callback with arguments added", () => {
  const mockCallback = jest.fn();
  doAdd(1, 2, mockCallback);
  //toHaveBeenCalledWith means call as arg 3
  expect(mockCallback).toHaveBeenCalledWith(3);
});

// xit('mock function with callback', () => {
//   const mockCallback = jest.fn(x => 42 + x);
//   forEach([0, 1], mockCallback);
//
//   // The mock function is called twice
//   expect(mockCallback.mock.calls.length).toBe(2);
//
//   // The first argument of the first call to the function was 0
//   expect(mockCallback.mock.calls[0][0]).toBe(0);
//
//   // The first argument of the second call to the function was 1
//   expect(mockCallback.mock.calls[1][0]).toBe(1);
//
//   // The return value of the first call to the function was 42
//   expect(mockCallback.mock.results[0].value).toBe(42);
// });
//
// function forEach(items, callback) {
//   for (let index = 0; index < items.length; index++) {
//     callback(items[index]);
//   }
//
// }
