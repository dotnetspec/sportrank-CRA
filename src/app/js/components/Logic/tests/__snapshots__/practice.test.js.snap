// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PRACTICE - test a functional component 1`] = `
<div>
  <h1>
    Hello React Function Component!
  </h1>
  <input
    onChange={[Function]}
    type="text"
    value="initial state"
  />
  <p>
    initial state
  </p>
</div>
`;
